peer_strategies.go

```diff
@@ -20,40 +20,57 @@
 
 package tchannel
 
-import (
-	"math"
-	"math/rand"
-	"time"
-)
+import "math"
 
-type scoreCalculator interface {
+// ScoreCalculator defines the interface to calculate the score.
+type ScoreCalculator interface {
 	GetScore(p *Peer) uint64
 }
 
-type randCalculator struct {
-	rng *rand.Rand
-}
+// ScoreCalculatorFunc is an adapter that allows functions to be used as ScoreCalculator
+type ScoreCalculatorFunc func(p *Peer) uint64
 
-func (r *randCalculator) GetScore(p *Peer) uint64 {
-	return uint64(r.rng.Int63())
+// GetScore calls the underlying function.
+func (f ScoreCalculatorFunc) GetScore(p *Peer) uint64 {
+	return f(p)
 }
 
-func newRandCalculator() *randCalculator {
-	return &randCalculator{rng: NewRand(time.Now().UnixNano())}
+type zeroCalculator struct{}
+
+func (zeroCalculator) GetScore(p *Peer) uint64 {
+	return 0
 }
 
-type preferIncoming struct {
-	rng *rand.Rand
+func newZeroCalculator() zeroCalculator {
+	return zeroCalculator{}
 }
 
-func (r *preferIncoming) GetScore(p *Peer) uint64 {
-	rn := uint64(r.rng.Int31())
-	if p.NumInbound() > 0 {
-		rn += math.MaxUint32
+type preferIncomingCalculator struct{}
+
+func (preferIncomingCalculator) GetScore(p *Peer) uint64 {
+	p.mut.RLock()
+	if p.NumInbound() <= 0 {
+		p.mut.RUnlock()
+		return math.MaxUint64
+	}
+	count := 0
+
+	for _, c := range p.outboundConnections {
+		c.outbound.mut.RLock()
+		count = count + len(c.outbound.exchanges)
+		c.outbound.mut.RUnlock()
+	}
+
+	for _, c := range p.inboundConnections {
+		c.outbound.mut.RLock()
+		count = count + len(c.outbound.exchanges)
+		c.outbound.mut.RUnlock()
 	}
-	return rn
+	p.mut.RUnlock()
+
+	return uint64(count)
 }
 
-func newPreferIncoming() *preferIncoming {
-	return &preferIncoming{rng: NewRand(time.Now().UnixNano())}
+func newPreferIncomingCalculator() preferIncomingCalculator {
```

Comment:
Add a comment about the strategy that this uses (prefers incoming, and then chooses based on least pending outbound calls)